test/celeritas/optical/ValidationUtils.hh

```diff
@@ -0,0 +1,136 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/ValidationUtils.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <array>
+#include <vector>
+
+#include "corecel/cont/Array.hh"
+#include "corecel/cont/Span.hh"
+#include "corecel/data/Collection.hh"
+#include "celeritas/UnitTypes.hh"
+#include "celeritas/grid/GenericGridData.hh"
+#include "celeritas/io/ImportPhysicsVector.hh"
+#include "celeritas/optical/MfpBuilder.hh"
+
+#include "Test.hh"
+#include "TestMacros.hh"
+
+#include "detail/ValidationUtilsImpl.hh"
+
+//! Equivalence macro for physics grids
+#define EXPECT_GRID_EQ(expected, actual) \
+    EXPECT_PRED_FORMAT2(::celeritas::testdetail::IsGridEq, expected, actual)
+
+//! Equivalence macro for physics tables (vectors of grids)
+#define EXPECT_TABLE_EQ(expected, actual) \
+    EXPECT_PRED_FORMAT2(::celeritas::testdetail::IsTableEq, expected, actual)
+
+namespace celeritas
+{
+namespace optical
+{
+namespace test
+{
+using namespace ::celeritas::test;
+//---------------------------------------------------------------------------//
+/*!
+ * Helper function to annotate units of a hard-coded test data array.
+ *
+ * Converts the arguments supplied in units \c UnitType to native units.
+ */
+template<class UnitType, class... Args>
+Array<real_type, sizeof...(Args)> constexpr native_array_from(Args const&... args)
```

Comment:
These are actually really cool. At some point (not this PR) we should use this in our other celeritas tests as well.